### PR TITLE
Make PolyMatrix arithmetics rigorous

### PR DESCRIPTION
**Pull request by sylee957**
#### Brief description of what is fixed or changed

1. PolyMatrix sanitizes string types of `ring` specification like `ring='ZZ'` to 

- MatrixShaping methods:
  - [x] `row_del`, `col_del` works and correctly inherits `ring`.
  - [x] `row_insert`, `col_insert` works and correctly unifies `ring`.
  - [x] `row_join`, `col_join` works and correctly unifies `ring`.
  - [x] `vec` works and correctly inherits `ring`.
  - [x] `row`, `col` works and correctly inherits `ring`.
  - [x] `extract` works and correctly inherits `ring`.
  - [x] `get_diag_blocks` works and correctly inherits `ring`.
  - [x] `hstack, vstack` works and correctly unifies `ring`.
  - [x] `reshape` works and correctly inherits `ring`.
  - [x] `diagonal` works and correctly inherits `ring`.
  - [ ] `vech`, `diag` works, but does not inherit the ring from given parameters.
  - [x] `tolist`, `todok` works without modification
- MatrixSpecial methods:
  - [x] `jordan_block` works and correctly inherits `ring` from Poly element given.
  - [x] `companion` works and correctly inherits `ring` from Poly element given.
  - [ ] `eye`, `ones`, `zeros` works, but ones and zeros are `S.One` and `S.Zero` than zeros corresponding to any `Poly`.
- MatrixProperties Methods
  - [x] `atoms`, `free_symbols`, `has`, `is_symbolic` works without modification.
  - [x] `is_square` works without modification.
  - [x] `is_symmetric`, `is_anti_symmetric` works.
  - [ ] `is_hermitian` is not implemented until #19531 is resolved.
  - [x] `is_diagonal` and `is_Identity` works.
  - [ ] `is_strongly_diagonally_dominant` and `is_weakly_diagonally_dominant` are not implemented because `Abs` does not work for `Poly`.
  - [x] `is_upper`, `is_lower`, `is_upper_hessenberg`, `is_lower_hessenberg` works.
  - [x] `is_zero_matrix` works.
- MatrixOperation Methods
  - [ ] `adjoint`, `conjugate` is not implemented until #19531 is resolved.
  - [ ] `as_real_imag`, `evalf`, `expand` will not be implemented because `Poly` does not have them.
  - [x] `transpose` works and correctly inherits `ring`.
  - [x] `rot90` works and correctly inherits `ring`.
  - [x] `permute_rows` and `permute_cols` works and correctly inherits `ring`.
  - [x] `applyfunc` works and correctly inherits `ring`, but it is not good to use for functions that does not give `Poly` output.
  - [x] `doit` works as an extension of `applyfunc`.
  - [ ] `replace`, `subs`, `xreplace` is not modified, because the issues with poly generators should be resolved from `Poly` side.
  - [ ] I've made `PolyMatrix.simplify` not to operate because Poly raises error when `simplify` is applied. However, `simplify(PolyMatrix)` may still raise error, but this is raised because PolyMatrix is converted to ImmutableMatrix in intermediate, and the only way to resolve this is introducing `ImmutablePolyMatrix`, so I'd skip this part.
  - [ ] I've marked `Abs` and `Mod` to be not implemented because the symbolic operators does not work for `Poly` objects properly. Although the modulo operation works as python operator like `PolyMatrix % n`, I see the operation can be confusing to the users because of the `auto=True` flag for the polynomials.
- MatrixDeterminant methods
  - [x] `det` supports `bareiss` and `berkowitz` methods.
  - [ ] `det` does not implement `lu` method yet.
  - [x] `charpoly` works.
- MatrixReduction methods
  - [x] `rref`, `echelon_form` works,
  - [x] `rank` works.
- MatrixEigen methods
  - [ ] `eigenvals`, `eigenvects`, `left_eigenvects` are not implemented.
  - [ ] `diagonalize`, `is_diagonalizable` are not implemented.
  - [ ] `bidiagonalize`, `bidiagonal_decomposition` are not implemented.
  - [ ] `jordan_form` is not implemented.
  - [ ] `singular_values` is not implemented.
- MatrixCalculus methods
  - [x] `diff`, `integral` works as formal derivative of polynomials.
  - [ ] `limit` is not implemented and `Poly` does not have it.
  - [ ] `jacobian` is not implemented for now, but it can be implemented as formal derivative of polynomials later.

TODO
- [ ] `adjugate`, `charpoly`, `cofactor`, `cofactor_matrix`, `det`, `minor`, `minor_submatrix`
- [ ] `echelon_form`, `is_echelon`, `rank`, `rref`, `elementary_row_op`, `elementary_col_op`
- [ ] `columnspace`, `nullspace`, `rowspace`, `orthogonalize`

#### Release Notes

- polys
  - Fixed `PolyMatrix` not unifying domains when addition, matrix multiplicaton is done.

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
There are lots of problems:
```julia
In [4]: from sympy.polys.polymatrix import PolyMatrix                                                                                                         

In [5]: xp = Poly(x)                                                                                                                                          

In [6]: yp = Poly(y)                                                                                                                                          

In [7]: M = PolyMatrix([[xp, yp], [yp, xp]])                                                                                                                  

In [8]: M.rref()                                                                                                                                              
Out[8]: 
⎛⎡            1               0⎤        ⎞
⎜⎢                             ⎥, (0, 1)⎟
⎝⎣Poly(0, x, y, domain='ZZ')  1⎦        ⎠

In [9]: M.zero                                                                                                                                                
Out[9]: 0

In [10]: M.zeros(2)                                                                                                                                           
Out[10]: 
⎡0  0⎤
⎢    ⎥
⎣0  0⎦

In [11]: M.eigenvals()                                                                                                                                        
---------------------------------------------------------------------------
TypeError 
~/current/sympy/sympy/sympy/simplify/simplify.py in simplify(expr, ratio, measure, rational, inverse, doit, **kwargs)
    571     # do deep simplification
    572     handled = Add, Mul, Pow, ExpBase
--> 573     expr = expr.replace(
    574         # here, checking for x.args is not enough because Basic has
    575         # args but Basic does not always play well with replace, e.g.

TypeError: replace() got an unexpected keyword argument 'simultaneous'
```
That last error comes up for many of the methods in `dir(M)`. It's due to:
```julia
In [12]: simplify(Poly(x))                                                                                                                                    
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-12-9e8504e61823> in <module>
----> 1 simplify(Poly(x))

~/current/sympy/sympy/sympy/simplify/simplify.py in simplify(expr, ratio, measure, rational, inverse, doit, **kwargs)
    571     # do deep simplification
    572     handled = Add, Mul, Pow, ExpBase
--> 573     expr = expr.replace(
    574         # here, checking for x.args is not enough because Basic has
    575         # args but Basic does not always play well with replace, e.g.

TypeError: replace() got an unexpected keyword argument 'simultaneous'
```
Also
```julia
In [15]: M.subs(x, y)                                                                                                                                         
Out[15]: 
⎡Poly(y, y, domain='ZZ')             y           ⎤
⎢                                                ⎥
⎣           y             Poly(y, y, domain='ZZ')⎦
```
That's due to the inconsistent behaviour of `Poly.subs`.

Also
```julia
In [17]: M.norm()                                                                                                                                             
/Users/enojb/current/sympy/sympy/sympy/core/operations.py:46: SymPyDeprecationWarning: 

Add/Mul with non-Expr args has been deprecated since SymPy 1.7. Use
Expr args instead. See https://github.com/sympy/sympy/issues/19445 for
more info.

  SymPyDeprecationWarning(
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-17-294197a3a567> in <module>
----> 1 M.norm()

~/current/sympy/sympy/sympy/matrices/matrices.py in norm(self, ord)
   1974                 ['f', 'fro', 'frobenius', 'vector']):
   1975                 # Reshape as vector and send back to norm function
-> 1976                 return self.vec().norm(ord=2)
   1977 
   1978             else:

~/current/sympy/sympy/sympy/matrices/matrices.py in norm(self, ord)
   1934         if self.rows == 1 or self.cols == 1:
   1935             if ord == 2 or ord is None:  # Common case sqrt(<x, x>)
-> 1936                 return sqrt(Add(*(abs(i) ** 2 for i in vals)))
   1937 
   1938             elif ord == 1:  # sum(abs(x))
...
~/current/sympy/sympy/sympy/core/mul.py in flatten(cls, seq)
    198             assert not a is S.One
    199             if not a.is_zero and a.is_Rational:
--> 200                 r, b = b.as_coeff_Mul()
    201                 if b.is_Add:
    202                     if r is not S.One:  # 2-arg hack

AttributeError: 'Poly' object has no attribute 'as_coeff_Mul'
```

Also:
```
M.log() - no attribute expand
M.one - Integer
M.ones - Integers
M.limit - no attribute limit
M.jacobian - data type not understood.
M.evalf()  - no attribute evalf
M.conjugate() - no attribute conjugate
M.cholesky - ditto
```

**Comment by oscarbenjamin** [comment]
I think it's a mistake to base the PolyMatrix class on Matrix. To the extent that methods are shared most will turn out to be broken or will get broken in future. It also makes it hard to improve the Matrix code if everything needs to written with the possibility that the elements might not be Expr.

With the DomainMatrix PR (#18844) I hope to remove the `RawMatrix` class which is the limitation on enforcing the args of `Matrix` to be `Basic`. If we can eliminate `PolyMatrix` as well or at least make it independent of `Matrix` then we can enforce the args of `Matrix` to be `Expr`.

**Comment by sylee957** [comment]
Unfortunately, I think that using `Poly` elements were the only way for using some capability of domain that `Expr` is not able to do, like finite field computations.
Why do you always insist that Expr to domain conversion should be the better way forward?

**Comment by oscarbenjamin** [comment]
I think that there can be a role for a `PolyMatrix` class but I think that it shouldn't be a subclass of Matrix. Also for speeding up general matrix computations it is better to start from the domain elements because using Poly there would add a level (actually multiple levels) that serve no purpose.

**Comment by sylee957** [comment]
However, even if the `PolyMatrix` is not inheriting `Matrix`, it should have similar API for shared methods like `add`, `multiply`, `pow`, `rref` `det`, and it is inherently not much different than not inheriting the whole class from MatrixBase or not. 
Unless if we are going to deprecate PolyMatrix.

Although `Poly` can be a heavier class than `Domain`, it is the best mathematical object to represent `Domain`.
So for the user API, `PolyMatrix` should be far more capable than `DomainMatrix`.

**Comment by oscarbenjamin** [comment]
> So for the user API, `PolyMatrix` should be far more capable than `DomainMatrix`.

I agree.

The DomainMatrix class should not be directed at users except for those users that are happy working with the domains as they are. We should use something like it to build higher-level classes analogous to Poly.

**Comment by oscarbenjamin** [comment]
To be clear I don't object to the changes here.

**Comment by sylee957** [comment]
I'm eventually willing to remove some trivial domain capability here like storing raw integers, and make every elements to be sympified to `Poly`, and also disallowing `EX` domain. 
But I'm keeping them for a while even if they are causing flawed behavior for untested stuff, because some of them were used by sympy codebase.

So I'd present some mock-ups of how user-level computations should be done here first by using Poly arithmetics, and I think that switching the Poly arithmetics to Domain arithmetics would not be that difficult than finding right domain objects from `Expr`.

**Comment by sylee957** [comment]
I can take a look at and would just mark all the stuff to raise `NotImplementedError` if they were not used by sympy internally, and have some flawed behavior.

Luckily I see determinant computations are working without much modifications.

**Comment by sylee957** [comment]
I'd only implemented basic methods like `det`, `rref` and marked every other method as not implemented because they are not working properly or mixing operations with `Expr`. 
And besides the `rref`, I don't think that any other methods were used across sympy codebase except for `prde.py` and I also think that `prde.py` is using PolyMatrix poorly because it mixes Poly with Expr,

So I've made the rref to fallback to the original one if there are `Expr` because they were somewhat working before, but I think that this clutter should be cleaned up so that `prde,py` is actually going to use `Poly` with `PolyMatrix` or going to use `Expr` with `Matrix`.

Since this PR can grow up too much if it is going to implement every other method that Matrix have that makes sense for PolyMatrix, I'd rather stop here and ask for reviews.

**Comment by oscarbenjamin** [comment]
I think this looks generally good. There are general improvements in the matrices code and a lot of things disabled or made to work properly for PolyMatrix.

It does seem a bit messy to duplicate some functions like berkowitz vector. Is there a cleaner way?

If this makes PolyMatrix generally usable then perhaps it should be documented somewhere. Is it already documented?

We should also deprecate using `Poly` in a normal `Matrix` with a warning that suggests using `PolyMatrix`.

**Comment by sylee957** [comment]
I also have concerns of having copy pasted procedures for matrix computations. 
Because I have seen lots of people writing wrong procedures and often such mistakes are hard to notice unless a lot of tests are run, which can be avoided by using some shared procedures because old tests will fail if something is modified wrong.

So I have an idea like e90b57149941bff7eefdc571eb4e5b713182f8a0 to use a common configurable procedure. Although it can complicate a lot of code, I think that this is the only one way to use common procedure.

**Comment by sylee957** [comment]
But at least some other problems of having duplicate code can be avoided by making an instance method equivalent to `_new`.

File: sympy/matrices/common.py [hunk 1]
```diff
@@ -2379,7 +2389,7 @@ class MatrixArithmetic(MatrixRequired):
     def _eval_Abs(self):
         return self._new(self.rows, self.cols, lambda i, j: Abs(self[i, j]))
 
-    def _eval_add(self, other):
+    def _eval_matrix_add(self, other):
         return self._new(self.rows, self.cols,
```
Comment (jksuom) [inline sympy/matrices/common.py:2393]: I think that this is the place where the unified domain of `self` and `other` should be added as the `ring` parameter to `_new`?

**Comment by oscarbenjamin** [comment]
I'm trying to understand what these last changes do.

```julia
In [25]: Matrix([[x, y]])                                                                                                                      
Out[25]: Matrix([[x, y]])

In [26]: Matrix([[x, y]]).ring                                                                                                                 
Out[26]: EX

In [21]: PolyMatrix([[x, y]])                                                                                                                  
Out[21]: Matrix([[x, y]])

In [22]: PolyMatrix([[x, y]]).ring                                                                                                             
Out[22]: EX

In [23]: PolyMatrix([[Poly(x, x), Poly(y, x), Poly(z, z)]])                                                                                    
Out[23]: Matrix([[Poly(x, x, domain='ZZ'), Poly(y, x, domain='ZZ[y]'), Poly(z, z, domain='ZZ')]])

In [24]: PolyMatrix([[Poly(x, x), Poly(y, x), Poly(z, z)]]).ring                                                                               
Out[24]: ZZ[y][x,z]
```

So all `Matrix` classes will have a `ring` attribute which is `EX` for normal matrix. The objects stored internally as the elements by `Matrix` will still be `Expr` rather than `expressiondomain.Expression`. PolyMatrix sometimes has `Expr` elements in which case the ring is `EX` so this is basically the same as `Matrix`.

Otherwise the elements of `PolyMatrix` are `Poly` and can be heterogeneous having different domains and generators. The `ring` attribute of a `PolyMatrix` is a `PolynomialRing` with ground domain and generators as the unification of the domains and generators of the `Poly` elements. In other words it is the domain that would be used to represent the matrix elements as domain elements rather than as `Poly`.

I think that it would be better if `PolyMatrix` did not allow `Expr` elements and either required homogeneous Poly or did unification at instantiation. It would make more sense to store the elements internally as `PolyElement` and e.g. convert to `Poly` on access so `M[0, 0]` could construct a `Poly` on the fly. Matrix operations like multiplication can do unification once per matrix operation whereas having `Poly` as the elements means unifying at each elementary multiply or add of the matrix elements (e.g. `N**3` times in a matrix multiply). Then `PolyMatrix` could have `domain` and `gens` attributes like `Poly` does. It could also have a `.rep` attribute which could be a `DomainMatrix` representing the elements internally that would be used for actual calculations.

**Comment by oscarbenjamin** [comment]
What are you thinking for this PR now @sylee957 ?

**Comment by sylee957** [comment]
I think that this PR can be changed to accomodate DomainMatrix internally,
But I'm not sure that it is going to be an easy change because of the prde.py is using PolyMatrix improperly with `Expr` classes than `Poly`
I think that it should be identified whether `prde` can use DomainMatrix internally (if they are using polynomial entries), or they should be using ordinary Matrix if they are accommodating non-polynomial functions, and I lack the background to distinguish them.

**Comment by sylee957** [comment]
I think that #19567 can be a better direction if things are easier to begin from scratch, but I think that the usage cases for PolyMatrix, NewMatrix should be cleaned up any ways

**Comment by czgdp1807** [comment]
@sylee957 Are you still working on this? Please resolve the conflicts if so. Thanks.

**Comment by sylee957** [comment]
Closed with reference to #21441

**Pull request closed without merging**